### PR TITLE
 [processor/cumulativetodelta] Convert cumulative Histograms to delta temporality (#12423)

### PR DESCRIPTION
**Pull request by mistodon**
**Description:** Cumulative Histogram metrics will now be converted to delta temporality and are no longer skipped. (This means that consumers who currently rely on Histograms being skipped would need to ensure they are excluded via config.)

**Link to tracking Issue:** https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/12423

**Testing:** Extended unit tests to cover Histogram metrics.

**Documentation:** Updated README to reflect new capability.

### CODE REVIEW

File: processor/cumulativetodeltaprocessor/processor.go
Comment (TylerHelmuth) [inline processor/cumulativetodeltaprocessor/processor.go:125]: Let's merge this line and line 127, where `bucketIdentity.MetricField` is set.

File: processor/cumulativetodeltaprocessor/processor.go
Comment (TylerHelmuth) [inline processor/cumulativetodeltaprocessor/processor.go:117]: Can you split out the bucketIdentities generation into a helper function?

File: processor/cumulativetodeltaprocessor/processor.go
Comment (TylerHelmuth) [inline processor/cumulativetodeltaprocessor/processor.go:131]: Nice, the struct feels way better.

File: processor/cumulativetodeltaprocessor/processor.go
Comment (TylerHelmuth) [inline processor/cumulativetodeltaprocessor/processor.go:167]: Since this is only for HistogramDataPoints lets update to convey that.  Maybe `convertHistogramFloatValue`.  Same with `convertIntValue` below.

File: unreleased/issue-12423.yaml
Comment (TylerHelmuth) [inline unreleased/issue-12423.yaml:16]: I think as an "ease-in" measure, we shouldn't convert histograms by default just yet.  Some users may be relying on the fact that the processor is skipping histograms because they want their histograms to be cumulative.  Let's add a new config option to enable histogram conversion support.  If enabled, histograms will be subjected to all the filtering logic and then converted if they pass.  If disabled, no histograms will be converted.

**Review by TylerHelmuth — COMMENTED** [review]
Looks great, thanks for taking this on.

File: unreleased/issue-12423.yaml
Comment (mistodon) [reply unreleased/issue-12423.yaml:16]: That makes sense. We could possibly do this more generically as a list of enabled metric types. So it would default to `["sum"]`, but `"histogram"` could be added as well. Just so that this could be extended to other metric types in future without adding bespoke new config options for those.

Comment (TylerHelmuth) [reply unreleased/issue-12423.yaml:16]: Thinking about it some more I think a [feature gate](https://github.com/open-telemetry/opentelemetry-collector/tree/main/service/featuregate) is a better solution that a temporary config option.  Feature gates allow users to enable features without us having to maintain a new config option and then deprecate it in the future.  

Here is an example of how the hostmetrics receiver is using feature gates to handle some changes to the metrics it collects:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/hostmetricsreceiver#feature-gate-configurations
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/12105

We'd define a timeline like the hostmetrics receiver and then eventually remove the feature gate, making converting histograms a default functionality.

Comment (mistodon) [reply unreleased/issue-12423.yaml:16]: This makes a lot of sense to me. I committed an attempt at adding a feature flag - is this the right way to do it? Does any other part of the codebase need to know about the flag?

Comment (TylerHelmuth) [reply unreleased/issue-12423.yaml:16]: This looks like a proper implementation to me.  Were you able to confirm locally that the feature flag works as expected?

Comment (mistodon) [reply unreleased/issue-12423.yaml:16]: In the process of doing that now 🙂

Comment (TylerHelmuth) [reply unreleased/issue-12423.yaml:16]: Awesome!

**Review by TylerHelmuth — COMMENTED** [review]
Can you update the README with details about the feature flag? Let's plan to remove it and make its functionality default no sonnet than release 0.60.0.

**Comment by mistodon** [comment]
Done. I wasn't sure how specific you wanted me to be with the timeline, so please feel free to edit/ask me to edit it.

File: processor/cumulativetodeltaprocessor/README.md
Comment (TylerHelmuth) [inline processor/cumulativetodeltaprocessor/README.md:83]:
```suggestion
The **processor.cumulativetodeltaprocessor.EnableHistogramSupport** feature flag controls whether cumulative histograms delta conversion is supported or not. It is disabled by default, meaning histograms will not be modified by the processor.  If enabled, which histograms are converted is still subjected to the processor's include/exclude filtering.
```

**Review by TylerHelmuth — APPROVED** [review]
One nit, otherwise looks good.

**Comment by TylerHelmuth** [comment]
@codeboten ready for a second review.

**Comment by mistodon** [comment]
Unfortunately, we think we've found a bug with this implementation. We're planning to investigate more on Monday, but for now, it's probably a good idea not to merge this. Sorry for the abrupt delay.

**Comment by mistodon** [comment]
Apologies again - we took a pretty thorough look into the bug and managed to establish that the issue occurs before reaching cumulative-to-delta. And in fact, it's equally possible to trigger the bug with sums and not histograms.

I've opened this for review again. 🙏

**Comment by TylerHelmuth** [comment]
@dashpole ready for a second review

**Comment by mistodon** [comment]
> If applicable, can you open an issue to track the bug?

This is the bug in question, for those curious: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/12746 - though it's not related to this PR in any specific way.

**Comment by TylerHelmuth** [comment]
@open-telemetry/collector-contrib-approvers please review.

**Comment by TylerHelmuth** [comment]
@mistodon please resolve conflicts caused by the removal of the Deprecated `metrics` config.

**Comment by mistodon** [comment]
Done 🙂

**Pull request merged**
